src/wavefront-reporter.js

```diff
@@ -0,0 +1,172 @@
+  ],
+  histogram: [
+    h => ['.count', h.count],
+    h => ['.min', h.min],
+    h => ['.mean', h.mean()],
+    h => ['.max', h.max],
+    h => ['.stddev', h.stdDev()],
+    h => ['.p50', h.percentiles([0.5])[0.5]],
+    h => ['.p75', h.percentiles([0.75])[0.75]],
+    h => ['.p95', h.percentiles([0.95])[0.95]],
+    h => ['.p98', h.percentiles([0.98])[0.98]],
+    h => ['.p99', h.percentiles([0.99])[0.99]],
+    h => ['.p999', h.percentiles([0.999])[0.999]]
+  ]
+};
+
+/**
+ * Reporter for sending metrics using direct ingestion.
+ * This reporter requires a server and a token to report data
+ * directly to a Wavefront server.
+ */
+export default class WavefrontReporter extends ScheduledReporter {
+  constructor({
+    server,
+    token,
+    source = 'wavefront-javascript-metrics',
+    registry = null,
+    reportingInterval = 1,
+    tags = null,
+    enableRuntimeMetrics = false
+  }) {
+    super(registry);
+    this.server = server;
+    this.source = source;
+    this.batchSize = 10000;
+    this.wavefrontClient = new wavefrontSDK.WavefrontDirectClient({
+      server: server,
+      token: token,
+      batchSize: this.batchSize,
+      flushIntervalSeconds: reportingInterval
+    });
+    this.histogramGranularity = new Set();
+    this.flushCurrentHist = true;
+    this.tags = !utils.isEmpty(tags) ? tags : {};
+  }
+
+  /**
+   * Collect metrics from registry and report them to Wavefront.
+   * TODO: With option to include current minute bin of histogram.
+   */
+  report() {
+    // Get all the metrics
+    let allMetrics = this.getMetrics();
+    for (let metrics of [
+      allMetrics.counters,
+      allMetrics.gauges,
+      allMetrics.meters,
+      allMetrics.timers,
+      allMetrics.histograms
+    ]) {
+      if (metrics.length > 0) {
+        for (let m of metrics) {
+          let [metricName, tags] = this._decodeMetric(m);
+          // If Wavefront histogram, client send_distribution
+          if (m instanceof WavefrontHistogram) {
+            let distributions = m.getDistribution();
+            if (this.flushCurrentHist) {
+              distributions.push(m.getCurrentMinuteDistribution());
+            }
+            for (let dist of distributions) {
+              this.wavefrontClient.sendDistribution(
+                metricName,
+                dist.centroids,
+                this.histogramGranularity,
+                dist.timestamp,
+                this.source,
+                tags
+              );
+            }
+            continue;
+          }
+          // If metric, client send_metric
+          for (let f of metricsFunctionMap[m.type]) {
+            const [suffix, value] = f(m);
+            console.log(metricName + suffix + ': ' + value);
+            this.wavefrontClient.sendMetric(
+              metricName + suffix,
+              value,
+              Date.now(),
+              this.source,
+              tags
+            );
+          }
+        }
+      }
+    }
```

Comment:
This whole `for` loop is not very easy to read. Can we refactor this?

Reply:
I left some comments to help understand the logic. Essentially it's just doing a loop through all the metrics and send to the wavefront server respectively based on its type.